src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -57,49 +57,67 @@ public DiagnosticService([ImportMany] IEnumerable<Lazy<IAsynchronousOperationLis
 
         private void RaiseDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs args)
         {
+            Contract.ThrowIfNull(sender);
+            var source = (IDiagnosticUpdateSource)sender;
+
             var ev = _eventMap.GetEventHandlers<EventHandler<DiagnosticsUpdatedArgs>>(DiagnosticsUpdatedEventName);
-            if (ev.HasHandlers)
+            if (!RequireRunningEventTasks(source, ev))
```

Comment:
this has nothing to do with this change. sure we can have 2 interfaces. you can search code to see all implementation of IDiagnosticUpdateSource. 

since diagnostic service put everything in memory, if source can do it better than diagnostic service, they can claim they support it. and they can put in temporary storage, persistent storage, out of proc as they fit or even re-calculate when asked.